magefile.go

```diff
@@ -1562,6 +1565,353 @@ func (Integration) Single(ctx context.Context, testName string) error {
+		})
+		builder.WriteString(t.Render())
+		builder.WriteString("\n")
+	}
+
+	return builder.String(), nil
+}
+
+func askForVM() (runner.StateInstance, error) {
+	vms, instances, err := listVMs()
+	if err != nil {
+		fmt.Errorf("cannot list VMs: %w", err)
+	}
+	fmt.Fprintf(os.Stderr, vms)
+
+	if len(instances) == 1 {
+		fmt.Fprintln(os.Stderr, "There is only one VM, auto-selecting it")
+		return instances[0], nil
+	}
+
+	id := 0
+	fmt.Fprint(os.Stderr, "Instance number: ")
+	if _, err := fmt.Scanf("%d", &id); err != nil {
+		return runner.StateInstance{}, fmt.Errorf("could not read instance number: %w:", err)
+	}
+
+	if id >= len(instances) {
+		return runner.StateInstance{}, fmt.Errorf("Invalid Stack number, it must be between 0 and %d", len(instances)-1)
+	}
+
+	return instances[id], nil
+}
+
+func askForStack() (runner.Stack, error) {
+	mg.Deps(Integration.ListStacks)
+
+	state, err := readFrameworkState()
+	if err != nil {
+		return runner.Stack{}, fmt.Errorf("could not read state file: %w", err)
+	}
+
+	if len(state.Stacks) == 1 {
+		fmt.Println("There is only one Stack, auto-selecting it")
+		return state.Stacks[0], nil
+	}
+
+	id := 0
+	fmt.Print("Stack number: ")
+	if _, err := fmt.Scanf("%d", &id); err != nil {
+		return runner.Stack{}, fmt.Errorf("cannot read Stack number: %w", err)
+	}
+
+	if id >= len(state.Stacks) {
+		return runner.Stack{}, fmt.Errorf("Invalid Stack number, it must be between 0 and %d", len(state.Stacks)-1)
+	}
+
+	return state.Stacks[id], nil
+}
+
+func generateEnvFile(stack runner.Stack) error {
+	fileExists := true
+	stat, err := os.Stat("./env.sh")
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("cannot stat 'env.sh': %w", err)
+		}
+		fileExists = false
+	}
+
+	if fileExists {
+		bkpName := fmt.Sprintf("./env.sh-%d", rand.Int())
+		if err := os.Rename(stat.Name(), bkpName); err != nil {
+			return fmt.Errorf("cannot create backup: %w", err)
+		}
+		fmt.Printf("%q already existed, it was moved to %q\n", stat.Name(), bkpName)
+	}
+
+	f, err := os.Create("./env.sh")
+	if err != nil {
+		return fmt.Errorf("Could not create './env.sh': %w", err)
+	}
+	defer f.Close()
+
+	fmt.Fprintf(f, "export ELASTICSEARCH_HOST=\"%s\"\n", stack.Elasticsearch)
+	fmt.Fprintf(f, "export ELASTICSEARCH_USERNAME=\"%s\"\n", stack.Username)
+	fmt.Fprintf(f, "export ELASTICSEARCH_PASSWORD=\"%s\"\n", stack.Password)
+
+	fmt.Fprintf(f, "export KIBANA_HOST=\"%s\"\n", stack.Kibana)
+	fmt.Fprintf(f, "export KIBANA_USERNAME=\"%s\"\n", stack.Username)
+	fmt.Fprintf(f, "export KIBANA_PASSWORD=\"%s\"\n", stack.Password)
+
+	return nil
+}
+
+// PrintState prints details about cloud stacks and VMs
+func (Integration) PrintState(ctx context.Context) {
+	fmt.Println("Virtual Machines")
+	mg.Deps(Integration.ListVMs)
+	fmt.Print("\n\n")
+	fmt.Println("Cloud Stacks")
+	mg.Deps(Integration.ListStacks)
+}
+
+// ListVMs lists all VMs in a human readable form, including connection details
+func (Integration) ListVMs() error {
```

Comment:
How about just `mage integration:instances`? The framework calls them instances, doesn't always have to be a VM. In the case of mac in the future it will probably be a physical machine.